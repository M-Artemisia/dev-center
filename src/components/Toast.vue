<template>
  <div class="syn-toast" :class="{ shown: showToast }">
    <div class="syn-flex justify-space-between align-stretch">
      <div class="syn-flex align-center">
        <svg width="14" height="14" viewBox="0 0 24 24">
          <path
            fill="currentColor"
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7c-.39-.39-.39-1.02 0-1.41.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"
          />
        </svg>
        <span class="syn-body-1 syn-text-light-primary syn-mb-0 syn-ml-8">{{message}}</span>
      </div>
      <button
        class="syn-btn icon text medium theme-light"
        data-e2e="toast-close-button"
        @click="closeToast"
      >
        <svg width="20px" height="20px" viewBox="0 0 24 24">
          <path
            fill="currentColor"
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
          />
        </svg>
      </button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Toast',
  data() {
    return {
      message: '',
      showToast: false,
      timeout: null
    }
  },
  created() {
    this.$root.$on('show-toast', this.openToast)
    this.$root.$on('hide-toast', this.closeToast)
  },
  methods: {
    clearCloseTimeout() {
      if (this.timeout) {
        clearTimeout(this.timeout)
        this.timeout = null
      }
    },
    closeToast() {
      this.clearCloseTimeout()
      this.showToast = false
    },
    openToast(message) {
      this.message = message
      // Show Toast
      this.showToast = true
      // Set Toast Dismiss Delay
      this.timeout = setTimeout(this.closeToast, 4000)
    },
  }
}
</script>
