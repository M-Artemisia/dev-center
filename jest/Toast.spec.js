import { mount } from '@vue/test-utils'
import Toast from '@/components/Toast'

describe('Toast', () => {
  test('should default to hidden', () => {
    const wrapper = mount(Toast)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show toast if show event is emitted from root', () => {
    const wrapper = mount(Toast)
    wrapper.vm.$root.$emit('show-toast')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should hide toast if hide event is emitted from root', () => {
    const wrapper = mount(Toast)
    wrapper.vm.$root.$emit('show-toast')
    wrapper.vm.$root.$emit('hide-toast')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should hide toast if close button is clicked', () => {
    const wrapper = mount(Toast)
    wrapper.vm.$root.$emit('show-toast')
    wrapper.find('.syn-btn').trigger('click')
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show message provided in show-toast event', () => {
    const wrapper = mount(Toast)
    wrapper.vm.$root.$emit('show-toast', 'Hello, world!')
    expect(wrapper.html()).toMatchSnapshot()
  })
})
