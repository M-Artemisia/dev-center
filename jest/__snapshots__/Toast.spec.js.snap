// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should default to hidden 1`] = `
<div class="syn-toast">
  <div class="syn-flex justify-space-between align-stretch">
    <div class="syn-flex align-center"><svg width="14" height="14" viewBox="0 0 24 24">
        <path fill="currentColor" d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7c-.39-.39-.39-1.02 0-1.41.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"></path>
      </svg> <span class="syn-body-1 syn-text-light-primary syn-mb-0 syn-ml-8"></span></div> <button data-e2e="toast-close-button" class="syn-btn icon text medium theme-light"><svg width="20px" height="20px" viewBox="0 0 24 24">
        <path fill="currentColor" d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"></path>
      </svg></button>
  </div>
</div>
`;

exports[`Toast should hide toast if close button is clicked 1`] = `
<div class="syn-toast">
  <div class="syn-flex justify-space-between align-stretch">
    <div class="syn-flex align-center"><svg width="14" height="14" viewBox="0 0 24 24">
        <path fill="currentColor" d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7c-.39-.39-.39-1.02 0-1.41.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"></path>
      </svg> <span class="syn-body-1 syn-text-light-primary syn-mb-0 syn-ml-8"></span></div> <button data-e2e="toast-close-button" class="syn-btn icon text medium theme-light"><svg width="20px" height="20px" viewBox="0 0 24 24">
        <path fill="currentColor" d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"></path>
      </svg></button>
  </div>
</div>
`;

exports[`Toast should hide toast if hide event is emitted from root 1`] = `
<div class="syn-toast">
  <div class="syn-flex justify-space-between align-stretch">
    <div class="syn-flex align-center"><svg width="14" height="14" viewBox="0 0 24 24">
        <path fill="currentColor" d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7c-.39-.39-.39-1.02 0-1.41.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"></path>
      </svg> <span class="syn-body-1 syn-text-light-primary syn-mb-0 syn-ml-8"></span></div> <button data-e2e="toast-close-button" class="syn-btn icon text medium theme-light"><svg width="20px" height="20px" viewBox="0 0 24 24">
        <path fill="currentColor" d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"></path>
      </svg></button>
  </div>
</div>
`;

exports[`Toast should show message provided in show-toast event 1`] = `
<div class="syn-toast shown">
  <div class="syn-flex justify-space-between align-stretch">
    <div class="syn-flex align-center"><svg width="14" height="14" viewBox="0 0 24 24">
        <path fill="currentColor" d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7c-.39-.39-.39-1.02 0-1.41.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"></path>
      </svg> <span class="syn-body-1 syn-text-light-primary syn-mb-0 syn-ml-8">Hello, world!</span></div> <button data-e2e="toast-close-button" class="syn-btn icon text medium theme-light"><svg width="20px" height="20px" viewBox="0 0 24 24">
        <path fill="currentColor" d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"></path>
      </svg></button>
  </div>
</div>
`;

exports[`Toast should show toast if show event is emitted from root 1`] = `
<div class="syn-toast shown">
  <div class="syn-flex justify-space-between align-stretch">
    <div class="syn-flex align-center"><svg width="14" height="14" viewBox="0 0 24 24">
        <path fill="currentColor" d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zM9.29 16.29L5.7 12.7c-.39-.39-.39-1.02 0-1.41.39-.39 1.02-.39 1.41 0L10 14.17l6.88-6.88c.39-.39 1.02-.39 1.41 0 .39.39.39 1.02 0 1.41l-7.59 7.59c-.38.39-1.02.39-1.41 0z"></path>
      </svg> <span class="syn-body-1 syn-text-light-primary syn-mb-0 syn-ml-8"></span></div> <button data-e2e="toast-close-button" class="syn-btn icon text medium theme-light"><svg width="20px" height="20px" viewBox="0 0 24 24">
        <path fill="currentColor" d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"></path>
      </svg></button>
  </div>
</div>
`;
